Improve addrev and ghmerge tools and their README

DDvO opened the pull request:
Various tweaks to make `ghmerge` more usable:
    * make gcc (rather than clang-3.6) the default for rebuilding
    * take the (SSH version of the) repo URL from the database
    * print some more (and slightly improve) info what is going on
    * show the log of the commits be pushed (after `addrev` is called)

Extend `review-tools/README`, making clear that `ghmerge` calls `addrev`

BTW, @mattcaswell, thanks for the Committer Instructions I had received by email.
I suggest adapting also these such that it becomes clear that  `ghmerge` uses `addrev`.
For instance:

```
Example usage adding PR and reviewer info to the last two commits:
$ addrev -2 --prnum=3711 matt @levitte master..

Example usage adding PR and reviewer info to the last commit and performing a non-merge push:
$ ghmerge --nomerge 3711 matt @levitte 
```

paulidale commented on review-tools/ghmerge:
```diff
@@ -82,14 +88,19 @@ function cleanup {
+echo -n Press return to merge to $REL: ; read foo
 git checkout $REL
 if [ "$MERGE" == "yes" ] ; then
     git merge --no-commit --squash $WORK
```
Not directly related this, but I'd like to have something more akin to `git rebase -i --autosquash` here.  Being able the tune the squashing is the cause of the majority of times I don't use `ghmerge`.

DDvO replied:
Good point. 
Since you already approved this PR, I suggest you do this in a separate PR.
If you prefer, I can `s/git merge --no-commit --squash/git rebase -i --autosquash/` here.

DDvO replied:
Now I see that `-i --autosquash` should better be added to the existing `git rebase` variant.
I've done this and updated and slightly extended the README accordingly.

paulidale replied:
Thanks, here is best since there are a number of other changes underway.
I'll clear my review.

DDvO commented:
I've just added also the following info on `(git)addrev` in `review-tools/README`:

```
The tool accesses databases on www.openssl.org via HTTPS.
The environment variable 'https_proxy' can be used to connect via a proxy.
The transfer may take many seconds, in particular with the '--list' option.
```

paulidale commented on review-tools/ghmerge:
```diff
@@ -100,12 +111,12 @@ else
 
 while true ; do
-    echo -n "Enter YES to push or NO to abort: "
+    echo -n "Enter YES/yes to push or NO/no to abort: "
```
I'm not sure this change is all that useful.

BTW: `pick-to-branch` also has the same yes/no question.

DDvO replied:
I just wanted to make clear that the simpler-to-type lower-case variants can also be used.
How if I instead `s/YES/yes/` and `s/NO/no/` in this tool and in `review-tools/pick-to-branch`?

DDvO replied:
I've just implemented the latter.

paulidale replied:
That's a better suggestion.  I end up typing just _y_ usually.
The intent of the question was to stop accidental continuation, but I do answer _no_ from time to time.

levitte commented:
While you're tweaking, would it be possible to have `$MERGE` default to "no"?  The mere fact that this script defaults to squashing all commits into one has had me avoid it like the plague.

paulidale commented:
A strong "yes" to defaulting `$MERGE` to no.

mspncp commented on review-tools/ghmerge:
```diff
@@ -59,9 +60,14 @@ set -- `python -c '
+import json, sys;
+print(str(json.load(sys.stdin)["head"]["repo"]["ssh_url"]))' </tmp/gh$$`
+REPO=$1
+rm /tmp/gh$$
```
Why not make it a single python call which loads the json file once and prints all three arguments?

DDvO replied:
I was thinking the same when I copied the Python code for simplicity.
I'm not a Python programmer so far, but meanwhile found how to do it nicely.

DDvO commented:
> While you're tweaking, would it be possible to have `$MERGE` default to "no"? The mere fact that this script defaults to squashing all commits into one has had me avoid it like the plague.

I was considering that as well but was reluctant because of the name of the tool.
Yet since there is strong consensus on changing the default I've done it now.

DDvO commented:
> A strong "yes" to defaulting `$MERGE` to no.

How about changing the name of the tool to `ghpush`?
This would reflect its semantics better, covering both the merge and rebase case.

mspncp commented on review-tools/ghmerge:
```diff
@@ -59,14 +59,12 @@ curl -s https://api.github.com/repos/openssl/$WHAT/pulls/$PRNUM >/tmp/gh$$
-print(str(json.load(sys.stdin)["head"]["label"]).replace(":", " "))' </tmp/gh$$`
+input = json.load(sys.stdin)
+print(str(input["head"]["label"]).replace(":", " "),
+      str(input["head"]["repo"]["ssh_url"]))'        </tmp/gh3531`
```
This is probably a leftover from testing? `/tmp/gh3531`

DDvO replied:
oops, yes! Fixed.

mspncp commented:
> > A strong "yes" to defaulting `$MERGE` to no.
> 
> How about changing the name of the tool to `ghpush`?
> This would reflect its semantics better, covering both the merge and rebase case.

I would leave the name as it is, because the main action is the merging (in the sense of integration) of the pull request into the master branch. That we integrate it linearly without creating an explicit merge commit is more a technical detail IMHO.

mspncp commented:
Essentially, it's a `git merge --ff-only`.

levitte commented on review-tools/README:
```diff
@@ -115,6 +115,9 @@ else 'git merge --no-commit --squash'.
 It works on the current branch, which should be master or one of the stable
 releases. For the branch to be merged a suitable remote must be available.
 
+The tool accesses databases on api.github.com and www.openssl.org via HTTPS.
```
www.openssl.org is incorrect.

DDvO replied:
So should I just remove "and www.openssl.org", 
or which other URLs are (potentially) contacted besides api.github.org?

mspncp replied:
Why not just write

    The tool accesses external databases via HTTPS (e.g., api.github.com).

since your main point is to inform the reader about the proxy settings?

DDvO replied:
Done

mspncp commented:
> > > A strong "yes" to defaulting `$MERGE` to no.
> > 
> > 
> > How about changing the name of the tool to `ghpush`?
> > This would reflect its semantics better, covering both the merge and rebase case.
> 
> I would leave the name as it is, because the main action is the merging (in the sense of integration) of the pull request into the master branch. That we integrate it linearly without creating an explicit merge commit is more a technical detail IMHO.

Looking at it once more, I think it's `$MERGE` which is misnamed, not `ghmerge`. I would suggest to name it `$SQUASH` and let it default to false.

DDvO commented:
> Looking at it once more, I think it's `$MERGE` which is misnamed, not `ghmerge`. I would suggest to name it `$SQUASH` and let it default to false.

I'm not so sure on this, since in both cases some (auto)squashing is done: 
`git merge --no-commit --squash $WORK`
vs.
`git rebase -i --autosquash`

DDvO commented:
Doing my most recent push using the new version of the tool,
I found it makes more sense to do the `git rebase -i --autosquash` before  `addrev` is called.
I've changed this and also added an initial `git pull` on master.

mspncp commented:
> > Looking at it once more, I think it's `$MERGE` which is misnamed, not `ghmerge`. I would suggest to name it `$SQUASH` and let it default to false.
> 
> I'm not so sure on this, since in both cases some (auto)squashing is done:
> `git merge --no-commit --squash $WORK`
> vs.
> `git rebase -i --autosquash`

It's a matter of how you view it: If the pull request has seven commits, each of which has several fixups, then the latter does a dumb squash-all-in-a-big-blob, while the former leaves the seven commits intact and only incorparates the fixups. But it's you to choose a name which suits you.

levitte commented:
> If the pull request has seven commits, each of which has several fixups, then the latter does a dumb squash-all-in-a-big-blob, while the former leaves the seven commits intact and only incorparates the fixups.

... but the other way around.  `--squash` is the "dumb squash-all-in-a-big-blob", while `--autosquash` "leaves the seven commits intact and only incorparates the fixups."

mspncp commented:
>... but the other way around.

Yeah, thanks for noticing. I mixed up the order.

DDvO commented:
> It's a matter of how you view it: If the pull request has seven commits, each of which has several fixups, then the former does a dumb squash-all-in-a-big-blob, while the latter leaves the seven commits intact and only incorporates the fixups.

[in this quote I've fixed the confusion mentioned above]

I know.

> But it's up to you to choose a name which suits you.

All right, so I've kept the variable name `MERGE`
but renamed the option '-nosquash' to the meanwhile more to-the-point '-autosquash'.

DDvO commented:
> Please reword the commit message of [e738545](https://github.com/openssl/tools/commit/e738545692ec2f960dbbfcaa09bb236eb9770138) to add an empty line between 'various tweaks' and the subitems

Done.

Unfortunately, such a change breaks the autosquash capability of the fixups for this commit because git references the base commit by its title (and not, e.g. by its hash-based ID).

In the future I'll make sure to put an empty line after the first line of multi-line commit messages.

mspncp commented:
>  git references the base commit by its title (and not, e.g. by its hash-based ID).

GIt does not reference the commit hash by design. Just imagine that git would do that. Then it would have to rewrite the commit messages of all fixups every time you rebase.
Fortunately, it's not such a big deal to reorder and edit the rebase-todo list manually, so an interactive rebase is still feasible :-)

levitte commented on review-tools/README:
```diff
@@ -101,16 +105,37 @@ Checkout branch for query 145:
+--nobuild         Do not use 'opensslbuild'.
+                  Else it is invoked with $CC defaulting to "ccache gcc".
+--tools           Select the 'tools' repository rather than 'openssl'.
+                  This implies '--merge' and '--nobuild'.
```
I would rather have `--autosquash` the default *everywhere*, at the very least for consistency.

paulidale replied:
Agreed.  With almost everyone using fixup commits, it makes sense.

DDvO replied:
Since `--autosquash` is, as written in the above addition to the README, an alias to `--nomerge` and `--nomerge` has recently become the default according to this PR, `--autosqash` is already the default.
Note also that for the alternative approach (triggered by `--merge` or  `--squash`) using `git merge --squash`, the `--autosquash` option is not supported by git.

levitte replied:
> `--autosqash` is already the default.

Then please explain to me the `MERGE=yes` in the lines quoted below.

``` shell
    --tools)
        WHAT=tools ; MERGE=yes ; BUILD=no ; shift
        ;;
    --web)
        WHAT=web ; MERGE=yes ; BUILD=no ; shift
        ;;
```

DDvO replied:
Ah, now I understand what you meant. Right, whoever did those  `--tools` and `--web` options for some (strange) reason added `MERGE=yes` to them. 

I was already considering dropping `MERGE=yes` for them. Done now.

DDvO commented:
I've just added  `--ff-only` to the `git merge` invocations to be on the safe side.

mspncp commented on review-tools/ghmerge:
```diff
@@ -113,12 +113,12 @@ git log $REL..
 if [ "$MERGE" == "yes" ] ; then
     echo -n "Press return to merge --squash to $REL: "; read foo
-    git merge --no-commit --squash $WORK
+    git merge --ff-only --no-commit --squash $WORK
```
👍

richsalz commented:
I suggest you avoid aliases, you don't need more than one way to do things.

t8m commented:
> I suggest you avoid aliases, you don't need more than one way to do things.

Yeah, especially because this is not a public-facing tool that we would need to keep backwards compatible.

DDvO commented:
> I suggest you avoid aliases, you don't need more than one way to do things.

I fully agree, but I'm surprised @richsalz that this suggestion comes from you.
I just went through the history of that tool and it turns out that essentially you introducted it,
including the aliases.

DDvO commented:
Anyway, which option names should survive?
`--merge` and `--nomerge` ?
`--squash`  and `--autosquash` ?
Or maybe `--merge` and `--autosquash` ?
Or anything new?

richsalz commented:
Yes, I wrote the tool.  (I wrote most of the things in this repo :).  I don't recall the aliases, but since you're doing a big cleanup, remove them.  Folks can always not install the current version.

DDvO commented:
No problem, but is there any preference on the names that should be kept?

Here's a further suggestion: `--merge` (implying squash) and `--rebase`  (implying autosquash).

t8m commented:
The relevant difference is only the --autosquash vs the forced --squash. I am actually not sure the --squash is really useful.

t8m commented:
It would be useful only if the PR's commit messages were really bogus and the committer would instead write a new message from scratch.
What might be useful is to have an option to skip the rebase -i --autosquash if the PR is clean and there are no fixup commits.

DDvO commented:
Thanks @t8m for sharing your thoughts.

So we could simply make autosquash the default,
`--squash` would become an alternative option, and all three other options can disappear.

DDvO commented:
Concerning the `rebase -i --autosquash`, in my view this does not hurt - the user will see the list of commits in his editor (which can be seen as opportunity for a sanity check) and if everything is nice and clean he can simply save&exit without changes.

mspncp commented:
> Thanks @t8m for sharing your thoughts.
> 
> So we could simply make autosquash the default,
> `--squash` would become an alternative option, and all three other options can disappear.

I agree with @t8m and @DDvO, but I would like to set the wording straight: The choice is not about the alternatives "autosquash" vs. "squash", it is about "interactive rebase" vs. "squash everything". 

The `--autosquash` option is just a comfort feature of the interactive rebase command. All git does is to reorder the items in the rebase-todo list and to replace some 'pick' actions by 'fixup' resp. 'squash' actions. So effectively it saves you some editing time, but the main point is that you are doing an interactive rebase.

And it makes a lot of sense that the interactive rebase is the default ghmerge action, whereas squashing requires an explicit `--squash` option.

DDvO commented:
Fine, so I've removed 3 merge/squash options and kept only non-interactive `--squash`
while interactive autosquash is the default.

DDvO commented:
I've meanwhile cleaned up up the (fixup) commits I've done so far.
All comments/requets have been addressed, also the formal change request by @mspncp.

DDvO commented:
I've done some further small improvements on `ghmerge` and its documentation.

I found that sometimes (after some more complex/messed commits that interfere with each other) the `--autosquash` option of interactive rebase re-arranges commits in a way that causes merge conflicts. Therefore I've fixed the cleanup behavior of `ghmerge` in this case and added the `--noautosquash` option that leaves out `--autosquash` on `git rebase -i`.

DDvO commented:
I've just added a further option to `ghmerge`:
using `--remote <arg>` one can select the git remote of the branch to pull from and push to.
This defaults to 'origin', the hitherto hard-wired value.

DDvO commented:
I've also slightly simplified the CLI of `addrev`:
Instead of `--prnum=NNN' one can now simply give `NNN`, e.g.:
```
addrev 1234 -2 steve @richsalz
```

mspncp commented:
> This defaults to 'origin', the hitherto hard-wired value.

Why not ask git instead of assuming 'origin'?

$ git remote -v | grep 'github.com/openssl/openssl'
github	https://github.com/openssl/openssl.git (fetch)
github	https://github.com/openssl/openssl.git (push)

mspncp commented:
Ups, wrong repo :-)

$ git remote -v | grep git.openssl.org
origin	git://git.openssl.org/openssl.git (fetch)
origin	openssl-git@git.openssl.org:openssl.git (push)

mspncp commented:
A classical use-case for good'ol awk:

$ git remote -v | awk '/git.openssl.org.*(push)/{ print $1; }'
origin

DDvO commented:
> $ git remote -v | awk '/git.openssl.org.*(push)/{ print $1; }'
> origin

@mspncp,  good thought and thanks for the code snippet. 
I've just extended it by ` | head -n 1` for the case that more than just one such remote is listed.
I've updated also the README in this PR accordingly. 
Again all your change requests are fulfilled.

DDvO commented:
Trying the `--squash` option of the latest `ghmerge` on #11134 yesterday revealed that its error handling still deserved some improvements, which in turn motivated a further tweak also on `addrev`. These are done in the two most recent commits.

DDvO commented:
Any further comments or requests for changes?
Else would be nice if this can approved such that it becomes directly usable.

DDvO commented:
Thanks @t8m and @paulidale for your swift approvals.

Can an OMC member please now do the merge?
You may already take advantage of the new `ghmerge` to merge itself :wink:, e.g., like this:
```
ghmerge --tools 59 @t8m paulidale
```

DDvO commented:
Thanks @t8m and @paulidale for your swift approvals.

Can an OMC member please now do the merge?
You may already take advantage of the new `ghmerge` to merge itself :wink:, e.g., like this:
```
ghmerge --tools 59 @t8m paulidale
```

paulidale commented:
I've merged it but I don't think I used the new ghmerge.  I would have used the new versions of things it calls.

t8m commented:
> I've merged it but I don't think I used the new ghmerge. I would have used the new versions of things it calls.

Well now after this merge the ghmerge is becoming really useful tool.

DDvO commented:
> Well now after this merge the ghmerge is becoming really useful tool.

Very pleased to hear!
